agent/php_user_instrument_hashmap.c

```diff
@@ -0,0 +1,315 @@
+  if (NULL != key->filename) {
+    zend_string_release(key->filename);
+  }
+  key->lineno = 0;
+}
+
+static inline size_t nr_zendfunc2bucketidx(size_t log2_num_buckets, zend_function* zf) {
+  /* default to lineno */
+  uint32_t hash = nr_php_zend_function_lineno(zf);
+
+  if (NULL != zf->op_array.function_name && !nr_zf_is_unnamed_closure(zf)) {
+    /* but use hash of function name when possible */
+    hash = ZSTR_HASH(zf->op_array.function_name);
+  } else if (NULL != zf->op_array.filename) {
+    /* and as last resort try hash of filename if available */
+    hash = ZSTR_HASH(zf->op_array.filename);
+  }
+
+  /* normalize to stay in-bound */
+  return (hash & ((1 << log2_num_buckets) - 1));
+}
+
+static inline bool zstr_equal(zend_string* zs1, zend_string* zs2) {
+  if (NULL == zs1 || NULL == zs2) {
+    return false;
+  }
+
+  if (ZSTR_LEN(zs1) != ZSTR_LEN(zs2)) {
+    return false;
+  }
+
+  if (ZSTR_HASH(zs1) != ZSTR_HASH(zs2)) {
+    return false;
+  }
+
+  if (0 != memcmp(ZSTR_VAL(zs1), ZSTR_VAL(zs2), ZSTR_LEN(zs1))) {
+    return false;
+  }
+
+  return true;
+}
+
+static bool nr_is_wraprec_for_zend_func(nr_php_wraprec_hashmap_key_t* key, zend_function* zf) {
+
+  if (nr_php_zend_function_lineno(zf) != key->lineno) {
+    return false;
+  }
+
+  if (zf->op_array.function_name && !nr_zf_is_unnamed_closure(zf)) {
+    if (!zstr_equal(key->function_name, zf->op_array.function_name)) {
+      return false;
+    }
+    /* compare scope if it is set */
+    if (zf->op_array.scope && !zstr_equal(key->scope_name, zf->op_array.scope->name)) {
+      return false;
+    }
+    return true;
+  }
+  
+  if (!zstr_equal(key->filename, zf->op_array.filename)) {
+      return false;
+  }
+
+  return true;
+}
+
+static int nr_php_wraprec_hashmap_fetch_internal(nr_php_wraprec_hashmap_t* hashmap,
+                             size_t hash_key, zend_function* zf, nr_wraprecs_bucket_t** bucket_ptr) {
+  nr_wraprecs_bucket_t* bucket;
+
+  for (bucket = hashmap->buckets[hash_key]; bucket; bucket = bucket->next) {
+    if (nr_is_wraprec_for_zend_func(&bucket->wraprec->key, zf)) {
+      *bucket_ptr = bucket;
+      return 1;
+    }
+  }
+
+  return 0;
+}
+
+static void nr_php_wraprec_hashmap_add_internal(nr_php_wraprec_hashmap_t* hashmap,
+                             size_t hash_key, nruserfn_t* wr) {
+  nr_wraprecs_bucket_t* bucket;
+
+  bucket = (nr_wraprecs_bucket_t*)nr_malloc(sizeof(nr_wraprecs_bucket_t));
+  bucket->prev = NULL;
+  bucket->next = hashmap->buckets[hash_key];
+  bucket->wraprec = wr;
+
+  if (hashmap->buckets[hash_key]) {
+    hashmap->buckets[hash_key]->prev = bucket;
+  }
+
+  hashmap->buckets[hash_key] = bucket;
+  ++hashmap->elements;
+}
+
+void nr_php_wraprec_hashmap_update(nr_php_wraprec_hashmap_t* hashmap, zend_function* zf, nruserfn_t* wr) {
+  nr_wraprecs_bucket_t* bucket = NULL;
+  size_t bucketidx;
+
+  if ((NULL == hashmap) || (NULL == zf) || (0 == wr)) {
+    return;
+  }
+
+  nr_php_wraprec_hashmap_key_set(&wr->key, zf);
```

Comment:
Do we need to worry about a case where this function fails to set the key? Is it even possible for that to happen?